### PR TITLE
el wu

### PR DESCRIPTION
**Pull request by envyaims**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making the changes.

### CODE REVIEW

File: solutions/cf-1017D.mdx [hunk 1]
```diff
@@ -0,0 +1,97 @@
+---
```
Comment (bqi343) [inline solutions/cf-1017D.mdx:1]: please add this to `solutions/gold`.

**Review by ryanchou-dev — APPROVED** [review]
nice!!

File: solutions/cf-1017D.mdx (hunk 1 above)
Comment (envyaims) [reply solutions/cf-1017D.mdx]: oopsies

File: solutions/cf-1017D.mdx [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+<CPPSection>
+  
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+const int INF = 1e9;
+
+int main(){
+	cin.tie(0) -> sync_with_stdio(0);
```
[the remark this answers is not in the document]
Comment (envyaims) [reply solutions/cf-1017D.mdx]: its needed

File: solutions/gold/cf-1017D.mdx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+<LanguageSection>
+
+<CPPSection>
+  
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+const int INF = 1e9;  // arbitrarily large integer
+
+int main(){
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1017D.mdx:29]: spaces

File: solutions/gold/cf-1017D.mdx [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+	cin.tie(0) -> sync_with_stdio(0);
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1017D.mdx:30]: remove if does fine w/o

File: solutions/gold/cf-1017D.mdx [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+	for (int i = 0; i < (1 << n); i++) {
+		for (int j = 0; j < n; j++) {
+			if (i & (1 << j)) {
+				wu[i] += w[j];
+			}
+		}
+	}
+	
+	// convert all strings to integers and store count
+	vector<int> str_cnt(1 << n);
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1017D.mdx:49]: rename to `str_count`

File: solutions/gold/cf-1017D.mdx [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+			if (s[j] == '1') {
+				val += 1 << j;
+			}
+		}
+		str_cnt[val]++;
+	}
+	
+	/*
+	 * precompute wu value between every pair of possible strings
+	 *.first stores the actual value, .second stores prefix of count
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1017D.mdx:63]:
```suggestion
	 * .first stores the actual value, .second stores prefix of count
```

File: solutions/gold/cf-1017D.mdx [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+			if (t[i] == '1') {
+				val += 1 << i;
+			}
+		}
+		// binary search for the least element with wu greater than k
+		auto it = upper_bound(pref_wu[val].begin(), pref_wu[val].end(), make_pair(k, INF));
+		if (it == pref_wu[val].begin()) {
+			cout << 0 << "\n";
+		} else {
+			cout << prev(it) -> second << "\n";
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1017D.mdx:92]:
```suggestion
			cout << prev(it)->second << "\n";
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/gold/cf-1017D.mdx (hunk 4 above)
Comment (envyaims) [reply solutions/gold/cf-1017D.mdx:30]: its not fine

File: solutions/gold/cf-1017D.mdx (hunk 5 above)
Comment (envyaims) [reply solutions/gold/cf-1017D.mdx:49]: bruh whats the point

Comment (SansPapyrus683) [reply solutions/gold/cf-1017D.mdx:49]: `cnt` is a horrible name

File: solutions/gold/cf-1017D.mdx [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+**Time Complexity:** $\mathcal{O}({2^n}^2 + qn)$
+
+<LanguageSection>
+
+<CPPSection>
+  
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+const int INF = 1e9;  // arbitrarily large integer
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1017D.mdx:27]: why not just `INT32_MAX`

File: solutions/gold/cf-1017D.mdx [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+source: CF
+title: The Wu
+author: Chongtian Ma
+---
+
+[Official Editorial (C++)](https://codeforces.com/blog/entry/61081)
+
+## Alternative Solution
+
+We can first precompute the wu values for each pair of strings. Then, we can prefix sum up the number of strings that gives each wu value of each string and utilize binary search. Note that in the implementation,
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1017D.mdx:12]: > Then, we can prefix sum up the number of strings that gives each wu value of each string

wha

Comment (envyaims) [reply solutions/gold/cf-1017D.mdx:12]: whats wrong with it

File: solutions/gold/cf-1017D.mdx (hunk 8 above)
Comment (envyaims) [reply solutions/gold/cf-1017D.mdx:27]: no reason, both are big enough

File: solutions/gold/cf-1017D.mdx (hunk 5 above)
Comment (envyaims) [reply solutions/gold/cf-1017D.mdx:49]: ho wcome

File: solutions/gold/cf-1017D.mdx (hunk 9 above)
Comment (SansPapyrus683) [reply solutions/gold/cf-1017D.mdx:12]: what does "prefix sum up" mean
we're constructing a pref sum array?
what are the values? what are the indices? what does computing a prefix sum represent? could you have an example?

File: solutions/gold/cf-1017D.mdx (hunk 8 above)
Comment (SansPapyrus683) [reply solutions/gold/cf-1017D.mdx:27]: can you use that instead

File: solutions/gold/cf-1017D.mdx (hunk 9 above)
Comment (envyaims) [reply solutions/gold/cf-1017D.mdx:12]: the array is literally named pref... "yeah we're doing prefix sums"

Comment (SansPapyrus683) [reply solutions/gold/cf-1017D.mdx:12]: i'd rather you put the explanation here than in the comments

Comment (envyaims) [reply solutions/gold/cf-1017D.mdx:12]: theres no point of reexplaining it through words when its directly demonstrated using code

Comment (SansPapyrus683) [reply solutions/gold/cf-1017D.mdx:12]: by that logic everything could just be in code comments

Comment (envyaims) [reply solutions/gold/cf-1017D.mdx:12]: its directly above the code block so people will know what is for, and what is the pref array. Having it before everything will just make it messier. Also, it's pretty brief and intuitive.

Comment (SansPapyrus683) [reply solutions/gold/cf-1017D.mdx:12]: then put something like "(further explained in the code)" here

File: solutions/gold/cf-1017D.mdx [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+	 */
+	vector<vector<pair<int, int>>> pref_wu(1 << n, vector<pair<int, int>>(1 << n));
+	for (int i = 0; i < (1 << n); i++) {
+		for (int j = 0; j < (1 << n); j++) {
+			pref_wu[i][j] = {wu.back() - wu[i ^ j], str_count[j]};
+		}
+		sort(pref_wu[i].begin(), pref_wu[i].end());
+		// accumulate prefix sums of count
+		for (int j = 1; j < (1 << n); j++) {
+			pref_wu[i][j].second += pref_wu[i][j-1].second;
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1017D.mdx:72]: spaces here

File: solutions/gold/cf-1017D.mdx (hunk 9 above)
Comment (envyaims) [reply solutions/gold/cf-1017D.mdx]: better?

**Pull request merged**